rfd/0083-machine-id-host-certs.md

```diff
@@ -0,0 +1,150 @@
+
+[`GenerateHostCert()`]: https://github.com/gravitational/teleport/blob/82c520c8183553f310459c3b4a96b70065ee268a/lib/auth/auth_with_roles.go#L2139
+[`CheckAccessToRule()`]: https://github.com/gravitational/teleport/blob/82c520c8183553f310459c3b4a96b70065ee268a/lib/services/role.go#L2309
+[Session Access]: https://github.com/gravitational/teleport/blob/82c520c8183553f310459c3b4a96b70065ee268a/lib/auth/session_access.go#L126
+
+### UX
+
+We have several options for issuing host certificates. Factors to consider
+include:
+1. Certificate lifespan. Do these certificates need to expire every hour?
+2. Does issuing them every hour create additional UX problems of its own?
+3. Does `sshd` gracefully reload these certificates, and if so, how can users be
+   expected to gracefully reload `sshd` when `tbot` is running as a user
+   process?
+
+Per (3), we have an open issue to help notify applications of changed
+certificates (https://github.com/gravitational/teleport/issues/11264). It's
+unclear if OpenSSH gracefully handles certificate rotations, though, and
+disconnecting sessions every 20 minutes would not be ideal.
+
+#### Option 0: No-op
+
+Users can already use bot identities to generate host certs with the `host_cert`
+permission granted, by passing the Machine ID `identity` file to `tctl`:
+
+```bash
+$ tctl -i path/to/identity auth sign \
+    --host=foo.example.com \
+    --format=openssh \
+    --out=myhost
+```
+
+This UX works fine in tandem with secure certificate issuance, but users are
+left to issue certs manually and to inform `sshd` of the change. If managed by
+cron / systemd timers / Ansible / etc, this could arguably be more 
+straightforward than `tbot` file change notifications (but complicates handling
+of CA rotation events).
+
+#### Option 1: Config template
+
+Config Templates in Machine ID are refreshed each iteration of the bots usual
+renewal loop (20min by default). This would be the most straightforward
+integration approach and we'd approach it like we do other special cert types
+for apps / databases / k8s clusters.
+
+Users would request a host cert via the config file:
+
+```yaml
+destinations:
+  - directory: /opt/machine-id
+    configs:
+      - host_cert:
+          principals: [foo.example.com]
+```
+
+While rendering the config template, the bot would call `GenerateHostCert()`,
+format it appropriately, and write it to disk. A new certificate would be 
+written every 20 minutes. 
+
+It would be up to the end user to inform sshd of the new certificate.
+
+#### Option 2: Out of band
+
+If we deem that these certs should be renewed at a different interval than other
+bot resources, we could add a new renewal loop for specifically these
+certificates. It would be additionally triggered on CA rotation, as regular
+renewals are today.
+
+Open questions: bots still renew all certs at startup. Do we maintain that
+behavior for these? If not, how do we know when to renew certs next?
```

Comment:
I suppose one option is to read the host certs and their issuance/expiry time out of them if the exist, and use that to determine a threshold after which we should renew them.

Reply:
Yeah, I think that's what would need to happen if we go this route.

I think my next step will be testing OpenSSH's behavior to see if we can just refresh its certs every 20 minutes without breaking things.

Reply:
closing the loop on this, there doesn't seem to be any real downside to frequent rotations:
* sshd reloads certs at login time, no explicit reload needed
  * even if we did need to reload sshd, ongoing sessions are still not impacted and keep running smoothly
* neither expired nor reloaded certs impact ongoing sessions, it's only ever checked at first connect
* clients starting new sessions automatically accept the new certs so long as they use our generated `known_hosts` which includes the cluster CA

All told, there's no real downsides to frequent cert rotation, though I did document a potential UX issue that could arise in the RFD. It's hopefully minor enough that we won't need to address it.